File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueMac.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.CamposToken;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Ataque ao MAC: tenta truncar o campo de integridade, zerá-lo e forçar
 * (força bruta de 1 byte) valores para ver se algum token adulterado é aceito.
 * Com um MAC de 32 bytes, nenhuma tentativa deveria passar.
 */
public class AtaqueMac implements AtaqueInterface {
    private final String mensagem;

    public AtaqueMac() {
        this("mensagem para o ataque de MAC");
    }

    public AtaqueMac(String mensagem) {
        this.mensagem = mensagem;
    }

    @Override
    public String nome() {
        return "Força bruta e truncamento do MAC";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        String token = alvo.encrypt(this.mensagem);
        byte[] decoded = alvo.base64urlDecode(token.substring(alvo.prefixo().length()));
        CamposToken campos = alvo.decompor(decoded);

        List<String> aceitos = new ArrayList<>();

        // 1) integridade zerada
        try {
            alvo.decrypt(this.montar(alvo, campos, new byte[campos.integridade.length]));
            aceitos.add("integridade zerada");
        } catch (Exception e) {
            // esperado
        }

        // 2) integridade truncada pela metade
        try {
            alvo.decrypt(this.montar(
                    alvo, campos, java.util.Arrays.copyOfRange(campos.integridade, 0, 16)));
            aceitos.add("integridade truncada (16 bytes)");
        } catch (Exception e) {
            // esperado
        }

        // 3) força bruta de 1 byte do MAC (255 variantes; pula o valor original,
        // que reconstruiria o próprio token válido e não é uma forja)
        byte[] base = campos.integridade.clone();
        int original = base[0] & 0xFF;
        for (int v = 0; v < 256; v++) {
            if (v == original) {
                continue;
            }
            byte[] tentativa = base.clone();
            tentativa[0] = (byte) v;
            try {
                alvo.decrypt(this.montar(alvo, campos, tentativa));
                aceitos.add("byte 0 do MAC = " + v);
            } catch (Exception e) {
                // esperado
            }
        }

        boolean vulneravel = !aceitos.isEmpty();

        Map<String, Object> dados = new HashMap<>();
        dados.put("aceitos", aceitos);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.CRITICA : Severidade.INFO,
                vulneravel
                        ? aceitos.size() + " variante(s) de MAC aceitas: "
                                + String.join("; ", aceitos.subList(0, Math.min(5, aceitos.size())))
                        : "Nenhuma das 257 variantes (zerada, truncada, 255 bytes forçados) foi aceita",
                dados);
    }

    private String montar(AlvoCriptografico alvo, CamposToken campos, byte[] integridade) {
        CamposToken alterado = new CamposToken(integridade, campos.ciphertext, campos.iv);
        return alvo.prefixo() + alvo.base64urlEncode(alvo.recompor(alterado));
    }
}

File: packages/go/internal/seguranca/ataques/ataque_mac.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// OpcoesMac configura o AtaqueMac.
type OpcoesMac struct {
	Mensagem string
}

// AtaqueMac tenta truncar o campo de integridade, zerá-lo e forçar (força bruta
// de 1 byte) valores para ver se algum token adulterado é aceito.
type AtaqueMac struct {
	mensagem string
}

// NovoAtaqueMac cria o ataque. Sem argumento, usa
// mensagem = "mensagem para o ataque de MAC".
func NovoAtaqueMac(opcoes ...OpcoesMac) *AtaqueMac {
	o := OpcoesMac{Mensagem: "mensagem para o ataque de MAC"}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueMac{mensagem: o.Mensagem}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueMac) Nome() string {
	return "Força bruta e truncamento do MAC"
}

// Executar roda o ataque.
func (a *AtaqueMac) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	token, err := alvo.Encrypt([]byte(a.mensagem))
	if err != nil {
		return seguranca.ResultadoAtaque{}, err
	}
	decoded, err := alvo.Base64urlDecode(token[len(alvo.Prefixo()):])
	if err != nil {
		return seguranca.ResultadoAtaque{}, err
	}
	campos := alvo.Decompor(decoded)

	montar := func(integridade []byte) string {
		return alvo.Prefixo() + alvo.Base64urlEncode(alvo.Recompor(seguranca.CamposToken{
			Integridade: integridade,
			Ciphertext:  campos.Ciphertext,
			IV:          campos.IV,
		}))
	}

	aceitos := []string{}

	// 1) integridade zerada
	if _, err := alvo.Decrypt(montar(make([]byte, len(campos.Integridade)))); err == nil {
		aceitos = append(aceitos, "integridade zerada")
	}

	// 2) integridade truncada pela metade
	if _, err := alvo.Decrypt(montar(campos.Integridade[:16])); err == nil {
		aceitos = append(aceitos, "integridade truncada (16 bytes)")
	}

	// 3) força bruta de 1 byte do MAC (255 variantes; pula o valor original,
	// que reconstruiria o próprio token válido e não é uma forja)
	base := append([]byte(nil), campos.Integridade...)
	for v := 0; v < 256; v++ {
		if byte(v) == base[0] {
			continue
		}
		tentativa := append([]byte(nil), base...)
		tentativa[0] = byte(v)
		if _, err := alvo.Decrypt(montar(tentativa)); err == nil {
			aceitos = append(aceitos, fmt.Sprintf("byte 0 do MAC = %d", v))
		}
	}

	vulneravel := len(aceitos) > 0
	severidade := seguranca.SevInfo
	detalhes := "Nenhuma das 257 variantes (zerada, truncada, 255 bytes forçados) foi aceita"
	if vulneravel {
		severidade = seguranca.SevCritica
		detalhes = fmt.Sprintf("%d variante(s) de MAC aceitas: %s", len(aceitos), strings.Join(aceitos[:min(5, len(aceitos))], "; "))
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes:   detalhes,
		Dados:      map[string]any{"aceitos": aceitos},
	}, nil
}
